pkg/cluster/cluster.go

```diff
@@ -82,63 +87,76 @@ type Cluster struct {
 	backupDir string
 }
 
-func New(c Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup) (*Cluster, error) {
-	return new(c, e, stopC, wg, true)
-}
-
-func Restore(c Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup) (*Cluster, error) {
-	return new(c, e, stopC, wg, false)
-}
-
-func new(config Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup, isNewCluster bool) (*Cluster, error) {
-	err := e.Spec.Validate()
-	if err != nil {
-		return nil, fmt.Errorf("invalid cluster spec: %v", err)
-	}
-
+func New(config Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup) *Cluster {
 	lg := logrus.WithField("pkg", "cluster").WithField("cluster-name", e.Name)
-	var bm *backupManager
-
-	if b := e.Spec.Backup; b != nil && b.MaxBackups > 0 {
-		bm, err = newBackupManager(config, e, lg, isNewCluster)
-		if err != nil {
-			return nil, err
-		}
-	}
 	c := &Cluster{
 		logger:  lg,
 		config:  config,
 		cluster: e,
 		eventCh: make(chan *clusterEvent, 100),
 		stopCh:  make(chan struct{}),
 		status:  &spec.ClusterStatus{},
-		bm:      bm,
+	}
+	err := c.setup(wg)
+	if err != nil {
+		c.logger.Errorf("fail to setup: %v", err)
+		if err != errClusterAlreadyFailed {
+			c.status.SetReason(err.Error())
+			c.reportFailedStatus()
+		}
+		return nil
+	}
+	return c
+}
+
+func (c *Cluster) setup(wg *sync.WaitGroup) error {
```

Comment:
can we move the line 156 -> 159 out of this function? also we do not need to pass in wg then.